Added another button at the top, a last button called Format on the Transform Dialogue

bayakarisa opened the pull request:
Partly Fixes #9721 
The structure of this button is very similar to numeric. There are 8 radio buttons. They are as follows:

Decimals
Scientific
Percent
NA
Zero
P-value
Fraction
Align

<img width="1858" height="975" alt="image" src="https://github.com/user-attachments/assets/2048173a-dfe1-43d4-87fa-09240d8b8a09" />

Decimals has an up-down from -10 to 22 with default 3.
Scientific has an up-down from 0 to 22 with default 4
Percent has an nup-down from 0 to 5 with default of 2
NA has a pull down into which you can optionally type what you want. Values in the pull down are na, NA, . (i.e. dot), --, missing 
Zero has a pull down where you can type if needed. In the pull down have 0, dot, -- and .
p-value this is a pull down with default p, also possible are * and p*.
Fractions has a pull down into which you can type a posuituve number. Default is 10 and you also have 20 and 100 there. 
Align has a pull down into which you can type. Values are dot, e, left, center, right.

@rdstern @lilyclements @Ag-Derek I have implemented what was discussed
I'll need help putting the R functions of the buttons

bayakarisa committed:
Added another button at the top, a last button called Format

The structure of this button is very similar to numeric. There are 8 radio buttons. They are as follows:

Decimals
Scientific
Percent
NA
Zero
P-value
Fraction
Align

lilyclements commented:
@bayakarisa this uses the R function `DescTools::Format`

E.g., if you want to see in R:

```r
?DescTools::Format
x <- pi * 10^(-10:10)

DescTools::Format(x, fmt="%")
```

* `<< variable name >>` - this is the name of the variable in the selector.
There are two options with a column receiver: You can either do 

```
ucrReceiver NAME.SetParameterIsString()
```

or

```
ucrReceiverNAME.SetParameterIsRFunction()
```

@bayakarisa why don't you try them both out, and see what happens in the R code for each one? (Hint: Print to script, and look at the `x = VARIABLE NAME` R code. It should be different depending if you set as string vs set as R function)

I've put in the issue some bits on the R code for the rest of the parameters and their values in this function

rdstern commented:
@bayakarisa just to add to @lilyclements comments:

a) Here is the "old" dialog:

<img width="942" height="978" alt="image" src="https://github.com/user-attachments/assets/cc0cc762-9cb4-4786-aaf1-f4019a610558" />

In comparison, you have made the dialog quite a lot wider, just to have the quite spacious radio buttons at the top.  I don't mind it a touch wider than before, but can you make the top radio buttons a bit smaller so it is not much changed from before?

Note that all the options are (I think) already working in the Transform keyboard of the calculator.  This should help you with the appropriate R code as the 2 ways should be consistent.  (And you can check they are working in Transform by trying them also in the calculator.

rdstern commented:
@bayakarisa I noite nothing has happened here for 2 weeks and there is no reply from you, on your plans.  Should this be passed to someone else?

bayakarisa committed:
Changes to the Format Menu

bayakarisa commented:
@rdstern 

1. Now only  the controls associated with a specific radio button (e.g., the NUD for decimals) are visible if that radio button is selected. 
<img width="1133" height="953" alt="Format" src="https://github.com/user-attachments/assets/36940a5f-6c50-4554-8e69-53888a17e827" />
2. I need help here with inserting the functionality of the radio buttons e.g.... 

This is the current script for  decimal :

```
Format(.x, digits=3)
data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
```

It should be like this 
```
size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
size1 <- round(size)
data_book$add_columns_to_data(data_name="survey", col_name="size1", col_data=size1, before=FALSE, adjacent_column="size")

data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
rm(list=c("size1", "size"))
```

lilyclements commented:
@bayakarisa 

1. I see you want to have this line:

```r
size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
```

This line can come automatically depending on how you set up the receiver type.  There are two options with a column receiver: You can either do

```r
ucrReceiverNAME.SetParameterIsString()
```

or

```
r
ucrReceiverNAME.SetParameterIsRFunction()
```

@bayakarisa why don't you try them both out, and see what happens in the R code for each one? (Hint: Print to script, and look at the x = VARIABLE NAME R code. It should be different depending if you set as string vs set as R function)


2. You're saying as well that you want to run like this:

```r
size1 <- round(size)
```

See [here](https://github.com/IDEMSInternational/R-Instat/issues/9721#issuecomment-3077988843) for the different options that you want to have. These are all in the function `Format` in the `DescTools` package.

So, for your first option, "decimals", you want to have

```r
size1 <- DescTools::Format(size, digits = VALUE IN NUD)
```

For this, you therefore want to have

a. You create a class, and set the function to be `Format`
b. For your class, you want the package name to be `DescTools`
c. For your class, you have two parameters here: First, `size` is the name of the variable in the receiver (check the parameter name for this)
d. Second parameter is `digits`, with the value in the associated NUD. 
e. You then want to assign this to the ucrSave.

bayakarisa committed:
Update dlgTransform.vb

bayakarisa commented:
@lilyclements I have Implemented what you asked 
Kindly check @berylwaswa 

1. I used the Survey data set  and used size 
<img width="1411" height="871" alt="Format1" src="https://github.com/user-attachments/assets/61b22b77-061f-45bf-8ec8-365e055b0f00" />
2. Here is the script that is generated it and  uses the DescTools::Format

<img width="1420" height="705" alt="Format2" src="https://github.com/user-attachments/assets/13472fdd-509a-435c-af53-f081b96040d2" />
3. Here is the result - size1 |
<img width="856" height="978" alt="Format3" src="https://github.com/user-attachments/assets/034e8335-71ff-48fb-a9be-0eb4acdc3a66" />

lilyclements committed:
Update dlgTransform.vb

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -88,9 +87,16 @@ Public Class dlgTransform
+    Private clsPipeOperator As New ROperator
+    Private clsTildaOperator As New ROperator
+    Private clsAssignOperator As New ROperator
+    
```
Here, I simply moved lines 109-117 up here so that all of our `As New RFunction`'s are together. This can make it easier for the next developer when looking for bits.

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -227,16 +223,11 @@ Public Class dlgTransform
-        ucrReceiverRank.SetParameter(New RParameter("x", 0))
-        ucrReceiverRank.SetParameterIsRFunction()
-
-
```
You set up ucrReceiverRank already (lines 164-169), so I've removed it from here so we don't have duplicated code

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -272,8 +263,6 @@ Public Class dlgTransform
 
-
-
         ucrPnlNonNegative.AddToLinkedControls(ucrInputPower, {rdoPower}, bNewLinkedHideIfParameterMissing:=True)
```
Just tidying up here by removing additional lines that are empty. I agree having a break of 1 line is nice and neat and makes things readable between different parts, but, can you go through and ensure there is never a break greater than 1 throughout?

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -774,11 +763,7 @@ Public Class dlgTransform
         ucrNudSignifDigits.AddAdditionalCodeParameterPair(clsSignifColsFunction, New RParameter("digits", 1), iAdditionalPairNo:=1)
         ucrInputPower.AddAdditionalCodeParameterPair(clsPowerColsOperator, New RParameter("y", 1), iAdditionalPairNo:=1)
-
-        ucrReceiverRank.AddAdditionalCodeParameterPair(clsDescToolsFormatFunction, New RParameter("x", 0), iAdditionalPairNo:=0)
```
See next comment, but in that one, I am referring to this line

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -802,6 +787,7 @@ Public Class dlgTransform
         ucrReceiverRank.AddAdditionalCodeParameterPair(clsScaleMinColsFunction, New RParameter("x", 0), iAdditionalPairNo:=29)
         ucrReceiverRank.AddAdditionalCodeParameterPair(clsBooleanColsOperator, New RParameter("x", 0), iAdditionalPairNo:=30)
         ucrReceiverRank.AddAdditionalCodeParameterPair(clsIsNAColsFunction, New RParameter("x", 0), iAdditionalPairNo:=31)
+        ucrReceiverRank.AddAdditionalCodeParameterPair(clsDescToolsFormatFunction, New RParameter("x", 0), iAdditionalPairNo:=32)
```
You had run this before, but, with one very small issue, which resulted in you writing a huge "If" statement, when it could be done much simpler!

The issue? You put `iAdditionalPairNo` as `0`.

We just want to put it as the next number, and put it with it's "brothers" (the other `ucrReceiverRank.AddAdditionalCodeParameterPair`'s.

We can see on lines 787, we have `iAdditionalPairNo := 29`, 788 is `:=30`, 789 is `:=31`, so, we make this one `:=32`.
(Note that 1-28 are all in here too, but to avoid having to write all of them out, I have just highlighted the previous few!)

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -833,6 +819,7 @@ Public Class dlgTransform
         ucrSaveNew.AddAdditionalRCode(clsPreviewOperator, iAdditionalPairNo:=13)
         ucrSaveNew.AddAdditionalRCode(clsBooleanOperator, iAdditionalPairNo:=14)
         ucrSaveNew.AddAdditionalRCode(clsIsNAFunction, iAdditionalPairNo:=15)
+        ucrSaveNew.AddAdditionalRCode(clsDescToolsFormatFunction, iAdditionalPairNo:=16)
```
Exactly the same as above, but with `ucrSaveNew`. I've just moved your `ucrSaveNew.AddAdditionalRCode(clsDescToolsFormatFunction, ....` line to be with it's brothers
And, I've set `iAdditionalPairNo:=16` since it's the 16th one for`ucrSaveNew`.

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -858,6 +845,9 @@ Public Class dlgTransform
         ucrChkOmitNA.SetRCode(clsMeanFunction, bReset)
         ucrChkPreview.SetRCode(clsConstantDummyFunction, bReset)
 
+        ' Set up for our "Format" options:
+        ucrNudDecimalPlaces.SetRCode(clsDescToolsFormatFunction, bReset)
+
```
Now, I've only set this up for `ucrNudDecimalPlaces`, but you need to go through and set this up for the other controls in the Format which are related :) e..g, `ucrNudScientific`, `ucrNudPercent`, and so forth until `ucrInputAlignOperations`.

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRmFunc)
-
+                If rdoDecimalFormat.Checked Then ' We can remove this If bit too!
                 ElseIf rdoScientific.Checked Then
```
This is the fun bit, because I got to remove code! And you will soon too. But I wanted to explain what I've removed and why, so see above.

But, in short, we don't need any of this code here because it's all automatically set up nicely for us when we run the `SetRCode` sub correctly!

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    clsGetColFunc.AddParameter("data_name", Chr(34) & strDataFrame & Chr(34))
-                    clsGetColFunc.AddParameter("col_names", strColumn)
-                    clsGetColFunc.AddParameter("use_current_filter", "FALSE")
-                    clsGetColFunc.SetAssignTo(strTempVar)
```
clsGetColFunc is set up automatically by:

1. We run `ucrReceiverRank.SetParameterIsRFunction()` in `Initialise`
2. We then link this to our `Format` function in the `SetRCode` where we now run `.AddAdditionalCodeParameterPair` with `:= 32`

Which means, this can alllllll go!

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    Dim strNewVar As String = ucrSaveNew.GetText
-                    Dim clsFormatFunc As New RFunction
-                    clsFormatFunc.SetPackageName("DescTools")
-                    clsFormatFunc.SetRCommand("Format")
```
We set this first bit up in `SetDefaults` so we don't need to repeat it here

```r
                    clsFormatFunc.SetPackageName("DescTools")
                    clsFormatFunc.SetRCommand("Format")
```

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    clsFormatFunc.SetPackageName("DescTools")
-                    clsFormatFunc.SetRCommand("Format")
-                    clsFormatFunc.AddParameter("x", strTempVar, bIncludeArgumentName:=False)
-                    clsFormatFunc.AddParameter("digits", ucrNudDecimalPlaces.GetText)
```
This is set up automatically in the `SetRCode` sub where we have `ucrNudDecimalPlaces.SetRCode(...)`, so we don't need to do it manually here. Neat!

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    Dim clsFormatFunc As New RFunction
-                    clsFormatFunc.SetPackageName("DescTools")
-                    clsFormatFunc.SetRCommand("Format")
-                    clsFormatFunc.AddParameter("x", strTempVar, bIncludeArgumentName:=False)
```
This is linked to our Format function in the SetRCode where we now run .AddAdditionalCodeParameterPair with := 32

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    clsAddColFunc.AddParameter("col_data", strNewVar)
-                    clsAddColFunc.AddParameter("before", "FALSE")
-                    clsAddColFunc.AddParameter("adjacent_column", strColumn)
-
```
This is all done automatically in the ucrSave - the ucrSave knows to do all this already, and so we just simply needed to link it to our `Format` function, which we did when ran `ucrSave.AddAdditionalCodeParameterPair with := 16`

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    clsRmFunc.SetRCommand("rm")
-                    clsRmFunc.AddParameter("list", "c(" & Chr(34) & strNewVar & Chr(34) & ", " & Chr(34) & strTempVar & Chr(34) & ")", bIncludeArgumentName:=True)
-
-
```
This is always done automatically, so this is never needed

lilyclements commented on instat/dlgTransform.vb:
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
-                    ucrBase.clsRsyntax.SetBaseRFunction(clsFormatFunc)
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRemoveLabelsFunction)
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRmFunc)
-
```
These bits are good, but, it is always the same regardless of which rdo is checked, so we just move this to the end of our `If rdoFormat` statement

lilyclements reviewed (commented):
@bayakarisa this is looking good! I wrote a long review, then decided the easiest way was to illustrate in one place what the amendments to do are, and then you can apply them.

This took a while to look through, and so I hope it should take you a while too in a way. It's important to understand these bits, and so don't worry about being fast. Worry about understanding it. If you have questions, then ask.

You need to pull changes into your branch too before continuing otherwise you'll get errors. 

So, I've made changes and outlined them for the `Decimal Places` option, you can now take over and do it for the Significant Values` etc options
(Spoiler alert: For Significant Values, I think you just need to add a `.SetRCode` argument where I mentioned it, and remove the code in the If statement!

lilyclements committed:
Update dlgTransform.vb

Forgot to add in one line that I realised when reviewing. That's that we just add

```r
ucrBase.clsRsyntax.SetBaseRFunction(clsDescToolsFormatFunction)
```

at the end of our `If` statemnet for `rdoFormat` :)

bayakarisa committed:
recommendations made

bayakarisa committed:
Merge branch 'format' of https://github.com/bayakarisa/R-Instat into format

bayakarisa commented:
> @bayakarisa please what is the status on this and is it ready for review
@Ag-Derek I need you to step in here

I am getting the same script for Percent and decimal since the parameter for both is digits 
and fraction too 

1.this is the script when i put decimal size 3

size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
size1 <- DescTools::Format(x=size, na.form="Missing", zero.form="0", align="l")
data_book$add_columns_to_data(data_name="survey", col_name="size1", col_data=size1, before=FALSE, adjacent_column="size")

data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
rm(list=c("size1", "size"))


and here is the script when i put  fractions of 3 on size

size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
size1 <- DescTools::Format(digits=3, x=size, fmt=3, na.form="Missing", zero.form="0", align="l")
data_book$add_columns_to_data(data_name="survey", col_name="size1", col_data=size1, before=FALSE, adjacent_column="size")

data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
rm(list=c("size1", "size"))

Ag-Derek commented:
Looks like the `digits` is being called when you actually do the percent parameter, I have sent you a message on whatsapp let me know when you are available within that time

bayakarisa committed:
Refactor format options handling in dlgTransform

Added Functionality

bayakarisa committed:
Referenced UpdateControlStates()

lilyclements commented:
@Ag-Derek and @berylwaswa can you review this?

berylwaswa reviewed (changes requested):
@bayakarisa the radio buttons are all working fine without any errors.

However, I noticed when I first opened the dialog it looked like the image below

<img width="550" height="478" alt="instat_2UbBEXDlkY" src="https://github.com/user-attachments/assets/f193964a-2eb0-4486-b53c-7689067501bf" />

The decimal places seem to be set by default to 3. But when I click on reset it's changed to zero. And when I go to the next radio button and go back to the decimal radi button it's set to zero. So maybe check that and make sure when someone opens the dialog for the first time it's always set to zero.

@Ag-Derek please check on your end.

bayakarisa committed:
Merge branch 'master' into format

bayakarisa committed:
changes requested

bayakarisa commented:
@lilyclements , Kindly Check
All minimum developer tests have been completed (OK/TestOK, Reset, UI behaviour, and edge cases). No outstanding issues found. .

lilyclements reviewed (changes requested):
@bayakarisa the code is working great! Just a couple of minor items:

Can you rearrange the rdo's so 

1. They're equally spaced apart?
2. Can you also move them to the right a bit so they're not covering the group box line on the left hand side? Currently the box is cut off. 
3. They're all aligned to start at the same point

<img width="359" height="352" alt="image" src="https://github.com/user-attachments/assets/9dabd2ee-a459-45bd-9618-8980b58f6347" />

You can fix 1 and 3 easily by selecting them all and clicking these:

For 1 -

<img width="1636" height="787" alt="image" src="https://github.com/user-attachments/assets/11367577-f784-4778-90ec-cccc9d544f64" />

For 3 -

<img width="1400" height="741" alt="image" src="https://github.com/user-attachments/assets/dd3b6dc0-5e35-4254-b790-7fce9e62b071" />

Then for two, just selected them all and nudge them to the right a few pixels together. 


Can you also check the Test OK enabled again? For example, here OK is enabled. It shouldn't be enabled if the rdo is checked and the associated NUD is empty

<img width="829" height="719" alt="image" src="https://github.com/user-attachments/assets/afd18d14-0851-4423-b34a-efbd9fd3c40d" />

bayakarisa committed:
Rearring and the TestOK fix

berylwaswa reviewed (commented):
@bayakarisa it's looking good on my side. @rdstern please do a final review before this is merged. Thanks

bayakarisa committed:
Fix Format tab state persistence and parameter bugs in Transform dialog

bayakarisa committed:
Fix Format tab parameter bugs and state persistence in Transform dialog

bayakarisa committed:
Address rdstern review comments on Format dialog

bayakarisa committed:
translation scripts - I did not have them in this branch thats why translation is failing

bayakarisa committed:
"Update translations"

bayakarisa committed:
This reverts commit 842b34eb71084cded4dbb9c9abfee68b303df33b.

bayakarisa committed:
Merge branch 'master' into format

bayakarisa committed:
Merge branch 'master' into format

rdstern reviewed (dismissed):
@bayakarisa well done, I am approving!
@Vitalis95 over to you.  I hope this will be easy, but I'm not sure.

Vitalis95 reviewed (commented):
@bayakarisa , great work
some few comments then we merge

Vitalis95 commented on instat/dlgTransform.vb:
```diff
@@ -746,6 +839,8 @@ Public Class dlgTransform
         ucrInputMultiply.SetRCode(clsScaleMultiplyOperator, bReset)
         ucrChkAddConstant.SetRCode(clsConstantDummyFunction, bReset)
+        ucrPnlNumericOptions.SetRCode(clsNumericDummyFunction, bReset)
+        ucrPnlFormatOptions.SetRCode(clsFormatDummyFunction, bReset)
```
remove this, you are setting rcode twice for ucrPnlFormatOptions

Vitalis95 commented on instat/dlgTransform.vb:
```diff
@@ -756,72 +851,116 @@ Public Class dlgTransform
         SetRCodeForControls(True)
         TestOKEnabled()
         ucrChkEditPreview.Checked = False
+        rdoDecimalFormat.Checked = True
```
do we need this? Seems you have set dummy function and `rdoDecimalFormat` is the default selected option

Vitalis95 commented on instat/dlgTransform.vb:
```diff
@@ -756,72 +851,116 @@ Public Class dlgTransform
+        End If
+        Return strValue
+    End Function
+
```
Remove this sub, since you are not calling it anywhere.

Vitalis95 commented on instat/dlgTransform.vb:
```diff
@@ -997,6 +1150,17 @@ Public Class dlgTransform
+            ElseIf rdoZero.Checked Then
+            ElseIf rdoPvalue.Checked Then
+            ElseIf rdoFraction.Checked Then
+            ElseIf rdoFractions.Checked Then
```
It looks like this `If...ElseIf` block is empty and doesn't have any effect. Can it be removed?

Vitalis95 commented on instat/dlgTransform.vb:
```diff
@@ -1073,6 +1237,108 @@ Public Class dlgTransform
+            clsDescToolsFormatFunction.AddParameter("fmt", "function(x) instatExtras::frac_den(x, den=" & UcrNudFraction.GetText & ")")
+
+        ElseIf rdoFractions.Checked Then
+            clsFormatDummyFunction.AddParameter("check", "Fractions", iPosition:=0)
```
Could we specify `iPosition` explicitly for all newly added `AddParameter` calls in this PR ?

Vitalis95 commented on instat/dlgTransform.vb:
```diff
@@ -1073,6 +1237,108 @@ Public Class dlgTransform
+
+        ElseIf rdoZero.Checked Then
+            UcrInputZeroOperations.Enabled = True
+
```
Do we still need `UpdateControlStates()`? It looks like all of these controls are already linked in `Initialise` using `AddToLinkedControls`

bayakarisa committed:
Merge branch 'master' into format

bayakarisa committed:
made necessary changes in reference to vitalis reviw

lilyclements committed:
trigger compile check

bayakarisa committed:
fixes to the translation error
